### PR TITLE
Serializer of IIR

### PR DESCRIPTION
**Pull request by twicki**
# Currenty under self review

## Technical Description

This should allow for all the data stored in the Internal IR to be written to file and be restored from there. This should be the first step towards a Metaheuristic search algorithm for finetuning.

#### Issues:
- We do not have a StencilFunctionToTemporaryPass yet and are currently unable to handle stencil functions. The pass needs to be present for this to work
- The Globals API is not sorted when code-generating and therefore we get different outputs if we serialze or not
- We need to figure out a good API for passing serialized data

### CODE REVIEW

File: src/dawn/IIR/IIRSerializer.cpp [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+  auto protoReadAccesses = protoAcesses->mutable_readaccess();
+  for(auto IDExtentsPair : accesses->getReadAccesses()) {
+    proto::iir::Extents protoExtents;
+    iir::Extents e = IDExtentsPair.second;
+    for(auto extent : e.getExtents()) {
+      auto protoExtent = protoExtents.add_extents();
+      protoExtent->set_minus(extent.Minus);
+      protoExtent->set_plus(extent.Plus);
+    }
+    protoReadAccesses->insert({IDExtentsPair.first, protoExtents});
```
Comment (cosunae) [inline src/dawn/IIR/IIRSerializer.cpp]: There is something weird in the accesses. If I deserialized the data you gave me, I get the following: 
for example for the statement: 
```          float hori_adv_dr_v=eddlat * adv_dr_1d_v_i + eddlon * adv_dr_1d_v_j```
I get
```
          w: __local_hori_adv_dr_v_103{0,0,0,0,0,0}
          r: __local_hori_adv_dr_v_103{0,0,0,0,0,0}
          r: __local_hori_adv_dr_v_103{0,0,0,0,0,0}
          r: __local_hori_adv_dr_v_103{0,0,0,0,0,0}
          r: __local_hori_adv_dr_v_103{0,0,0,0,0,0}
```

File: src/dawn/Support/ASTSerialier.h [hunk 2]
```diff
@@ -0,0 +1,120 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
```
Comment (cosunae) [inline src/dawn/Support/ASTSerialier.h]: typo in the file name

File: src/dawn/Support/ASTSerialier.cpp [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+    currentExprProto_.pop();
+  }
+
+  setLocation(protoExpr->mutable_loc(), expr->getSourceLocation());
+}
+
+void ProtoStmtBuilder::visit(const std::shared_ptr<FieldAccessExpr>& expr) {
+  auto protoExpr = getCurrentExprProto()->mutable_field_access_expr();
+
+  protoExpr->set_name(expr->getName());
```
Comment (cosunae) [inline src/dawn/Support/ASTSerialier.cpp]: In the case of stencil function inline example you passed me of horizontal diffusion, 
the names stores are the names of the stencil function scope, while we need the original names. 
There is something wrong here

File: src/dawn/IIR/IIR.proto [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+// @brief The Protobuf description of all the required members to describe a MultiStage of the IIR
+//
+// A MultiStage usually corresponds to the outer loop (usually over k) of the loop nest. In CUDA
+// gridtools multistages reflect kernels.
+message MultiStage{
+  enum LoopOrder { Forward = 0; Backward = 1;  Parallel = 3;}
+    repeated Stage stages = 1;
+    LoopOrder looporder = 2;
+    int32 MulitStageID = 3;
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/IIR/IIR.proto]: typo fixed

File: src/dawn/IIR/IIR.proto [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    int32 stageID = 2;
+
+}
+
+// @brief The Protobuf description of all the required members to describe a MultiStage of the IIR
+//
+// A MultiStage usually corresponds to the outer loop (usually over k) of the loop nest. In CUDA
+// gridtools multistages reflect kernels.
+message MultiStage{
+  enum LoopOrder { Forward = 0; Backward = 1;  Parallel = 3;}
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/IIR/IIR.proto]: fixed

File: src/dawn/IIR/IIR.proto [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    Acesses calleeAccesses  = 3;
+}
+
+// @brief The Protobuf description of all the required members to describe a DoMethod of the IIR
+//
+// A Do-method is a collection of Statements with corresponding Accesses of a specific vertical
+// region
+message DoMethod{
+    repeated StatementAcessPair stmtaccesspairs = 1;
+    int32 DoMethodID = 2;
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/IIR/IIR.proto]: fixed throughout

File: src/dawn/IIR/IIR.proto [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    int32 minus = 1;
+    int32 plus = 2;
+}
+
+message Extents{
+    // the three dimensions of an access-extent [Extent<i>, Extent<j>, Extent<k>]
+    repeated Extent extents = 1;
+}
+
+message Acesses{
```
[the remark this answers is not in the document]
Comment (twicki) [reply src/dawn/IIR/IIR.proto]: fixed

File: src/dawn/Support/ASTSerialier.cpp (hunk 3 above)
Comment (twicki) [reply src/dawn/Support/ASTSerialier.cpp]: I'll investigate

File: src/dawn/IIR/IIRSerializer.cpp (hunk 1 above)
Comment (twicki) [reply src/dawn/IIR/IIRSerializer.cpp]: I'll look at it

File: src/dawn/Support/ASTSerialier.h (hunk 2 above)
Comment (twicki) [reply src/dawn/Support/ASTSerialier.h]: how you catch this is beyond me, fixed

**Comment by twicki** [comment]
depcirated due to newer version

**Pull request closed without merging**